### PR TITLE
Back up MySQL database

### LINKED ISSUE #354: Data versioning, backup and pipeline tracking for SQL databases
!This is an exploratory issue.

Questions:
* Given an existing database, what is the recommended way to back it up? Could it be tracked with DVC?
* Does the concept of versioning exist for databases? More specifically, is it possible to add a version tag and then check it out at any point?
* If the database is a result of taking multiple raw sources, preprocessing them and adding them as rows to existing tables would there be any benefits/issues with turning it into a DVC pipeline? See below an example
```text
  * initialize empty tables
  * take raw `data_1`
  * apply `preprocessing_1.sh` to `data_1` to get `preprocessed_data_1`
  * populate the database with `preprocessed_data_1` 
  * take raw `data_2`
  * apply `preprocessing_2.sh` to `data_2` to get `preprocessed_data_2`
  * populate the database with `preprocessed_data_2` 
  * ...
```

### PR DESCRIPTION
**Pull request by Stannislav**
Fixes #354 .

# Description
As commented in #354, a sensible approach is described [here](https://jpaulmora.github.io/2020/03/22/how-to-properly-backup-mysql/):
- Cron job dumps whole database to binary files using [`mysqldump`](https://dev.mysql.com/doc/mysql-backup-excerpt/8.0/en/using-mysqldump.html) daily/weekly.
- MySQL [binary log](https://dev.mysql.com/doc/refman/8.0/en/binary-log.html) records incremental changes to DB between file backups.

### In this PR:
- We add the script `make-backup` to the `mysql` docker image that can be run to create database dumps
- This script should be called via docker: `docker exec -t mysql-container make-backup MYSQL_USER MYSQL_PASSWORD DB_NAME`.
- The backup files will be placed into the `/backup` directory within the container. This directory is configured to be an anonymous local docker volume, but it's better to explicitly mount a volume into that directory via the `-v` parameter in `docker run`.
- We configure the binary logs to have an expiry of 7 days

### Not in this PR: cron job
We should try and configure an external cron job that calls the `docker exec` command above every 7 days if we want to make backups that way.

I experimented with having a cron service running within the mysql container, but
- it's not really recommended that a docker container does more than one thing / runs more than one process
- one would need to store the MySQL password on the container in order to be able to call `mysqldump`
- one would need to know upfront the names of the databases that will be backed up (or back up all databases)

If one is happy with the above drawbacks, then we can still go for that solution. Actually, the first commit in this PR (2b354da) already did all the necessary preparations apart from adding the actual cron job (e.g. just a script in `/etc/cron.weekly`):
- Installing the cron and syslog services
- Starting the cron and syslog services at container startup

# How to test?
Build a test image
```bash
docker build -f docker/mysql.Dockerfile -t my-test-mysql-image .
```

Start a container
```bash
mkdir mysql-backup
docker run \
--rm \
-d \
-e MYSQL_ROOT_PASSWORD=whatever \
-e MYSQL_DATABASE=testdb \
--name my-test-mysql-server \
-v $PWD/mysql-backup:/backup \
-P \
my-test-mysql-image
```
(normally one would probably also mount the `/var/lib/mysql` directory, but it's not necessary here)

Check if it's running via `docker ps`. If not try replacing `-d` by `-it` and checking what's wrong.

Populate the database with some random data:
```python
import random
import string

import sqlalchemy

engine = sqlalchemy.create_engine("mysql://root:whatever@<host>:<port>/testdb")
with engine.begin() as conn:
    conn.execute("drop table if exists mytable; create table mytable(name text, age int);")
    for _ in range(100):
        name = "".join(random.sample(string.ascii_letters, 10))
        age = random.randint(18, 90)
        conn.execute(f"insert into mytable(name, age) values ('{name}', {age})")
```

### Test the file backup
Trigger a backup 
```
export MYSQL_USER=root
export MYSQL_PWD=whatever
docker exec -t -e MYSQL_USER -e MYSQL_PWD my-test-mysql-server make-backup testdb
```

Check the backup is there:
```
ls mysql-backup
```

### Test recovery from backup
Drop the database
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'drop database testdb'
```

Check database doesn't exist
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
ERROR 1049 (42000) at line 1: Unknown database 'testdb'
```

Recover form backup file
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'create database testdb'
docker exec -i my-test-mysql-server mysql -uroot -p$MY_PWD testdb < /path/to/backup.sql
```

Check data is there
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
+----------+
| count(*) |
+----------+
|      100 |
+----------+
```

### Test recovery from binlog files
More info here: https://dev.mysql.com/doc/refman/8.0/en/point-in-time-recovery.html

Check binlog is enabled
```
docker exec -it my-test-mysql-server mysql -uroot -p -e "show variables like 'log_bin'"
Enter password: 
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| log_bin       | ON    |
+---------------+-------+
```

Check data is there
```
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
+----------+
| count(*) |
+----------+
|      100 |
+----------+
```

Get current position in the binlog
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'show master status'
Enter password: 
+---------------+----------+--------------+------------------+-------------------+
| File          | Position | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+---------------+----------+--------------+------------------+-------------------+
| binlog.000002 |    15473 |              |                  |                   |
+---------------+----------+--------------+------------------+-------------------+
```

Drop the database
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'drop database testdb'
```

Check database doesn't exist
```bash
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
ERROR 1049 (42000) at line 1: Unknown database 'testdb'
```

Replay all binlog records up to the point where we dropped the database
```bash
docker exec -it my-test-mysql-server bash
cd /var/lib/mysql/
mysqlbinlog --stop-position 15473 binlog.000001 binlog.000002 | mysql -uroot -p
exit
```

Check data is there
```
docker exec -it my-test-mysql-server mysql -uroot -p -e 'select count(*) from testdb.mytable'
Enter password: 
+----------+
| count(*) |
+----------+
|      100 |
+----------+
```

# Checklist

- [x] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
  (if it is not the case, please create an issue first).
- [x] Unit tests added.
  (if needed)
- [ ] Documentation and `whatsnew.rst` updated.
  (if needed)
- [x] `setup.py` and `requirements.txt` updated with new dependencies.
  (if needed)
- [x] Type annotations added.
  (if a function is added or modified)
- [x] All CI tests pass.

### CODE REVIEW

**Comment by Stannislav** [comment]
One might be worried about the following warnings in the log during the server startup:
```text
[Warning] [MY-010161] [Server] You need to use --log-bin to make --binlog-expire-logs-seconds work.
[Warning] [MY-010453] [Server] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
```

Contrary to the warning binary logs seem to be enabled as one can check via
```
docker exec -it bbs-mysql mysql -uroot -p -e "show variables like 'log_bin'"
Enter password: 
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| log_bin       | ON    |
+---------------+-------+
```

The second warning is a non-issue as per [here](https://github.com/docker-library/mysql/issues/307#issuecomment-313784886)

**Review by jankrepl — COMMENTED** [review]
Thanks for the PR and the interesting description!

I left a few comments in the code.

File: docker/mysql.Dockerfile [hunk 1]
```diff
@@ -15,17 +15,30 @@
...
+apt-get clean &&\
+rm -r /var/lib/apt/lists/*
+
+# Limit incremental binary log to 7 days. This is a system variable and has to
+# go in the [mysqld] section, which is in docker.cnf
+# Accordingly it would make sense to do file dumps every 7 days
+RUN echo "binlog_expire_logs_seconds = 604800" >> /etc/mysql/conf.d/docker.cnf
+
+# Install the backup script
+COPY docker/mysql-make-backup /usr/local/bin/make-backup
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply docker/mysql.Dockerfile:39]: Yes, exactly, but I think that's what we do with all other docker images. For other images we need the source/data files, I guess that's the main reason why we do that. For this one it's not really necessary, but I just went for consistency... What do think?

[the remark this answers is not in the document]
Comment (jankrepl) [reply docker/mysql.Dockerfile:39]: If this is consistent with what we do elsewhere let's keep it the way it is:))

File: docker/mysql-make-backup [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+
+MYSQL_USER="$1"
+export MYSQL_PWD="$2"
```
[the remark this answers is not in the document]
Comment (Stannislav) [reply docker/mysql-make-backup:4]: Yes that's true.

One could maybe argue that we can create a specific user just for backing up that would only have read access to the relevant databases. Then maybe it wouldn't be so critical.

I think there's also an alternative: the scripts reads the value from the environment. Then the callee would have to do something like that:
```bash
export MYSQL_USER=root
export MYSQL_PWD=whatever
docker exec -t -e MYSQL_PWD -e MYSQL_USER my-test-mysql-server make-backup testdb
```
What do you think?

[the remark this answers is not in the document]
Comment (jankrepl) [reply docker/mysql-make-backup:4]: I think that I would lean towards the environment variable solution.

[the remark this answers is not in the document]
Comment (Stannislav) [reply docker/mysql-make-backup]: Implemented in c96d9ce

[the remark this answers is not in the document]
Comment (jankrepl) [reply docker/mysql-make-backup]: Great! Thank you!

**Review by EmilieDel — COMMENTED** [review]
Thanks for the work and very nice description. 

So, the backup is done through 2 steps: the file backup and the binary logs one. The second one is set up but the first one not yet. Did I understand correctly ?

**Comment by Stannislav** [comment]
> Thanks for the work and very nice description.
> 
> So, the backup is done through 2 steps: the file backup and the binary logs one. The second one is set up but the first one not yet. Did I understand correctly ?

Yes, pretty much. The binary logs work out of the box. For the file back ups the code/scripts have been set up, but they won't happen automaticaly - for this we need to set up a cron job somewhere. But they can already be triggered automatically.

**Review by pafonta — COMMENTED** [review]
Thanks for the investigations about backing up a MySQL database.

The PR description says that it fixes the issue #354. However, the issue has a broader scope. Especially, the issue originated from the discussion about the incremental population of the database.

Then, I understand that the PR brings a solution to how to back up the database. But I don't see a mention of how it helps about being able to populate the database incrementally.

**Is it expected? Was there an undocumented shift of goal at some point?**

**Comment by Stannislav** [comment]
> Thanks for the investigations about backing up a MySQL database.
> 
> The PR description says that it fixes the issue #354. However, the issue has a broader scope. Especially, the issue originated from the discussion about the incremental population of the database.
> 
> Then, I understand that the PR brings a solution to how to back up the database. But I don't see a mention of how it helps about being able to populate the database incrementally.
> 
> **Is it expected? Was there an undocumented shift of goal at some point?**

As discussed on zoom, further issues like incremental population of the database will be addressed in new issues/PRs.

**Pull request merged**